workspace/src/path_planning/waypoints_path_planner/waypoints_path_planner/waypoints_path_planner.py

```diff
@@ -0,0 +1,221 @@
+
+    Attributes:
+        lookahead: How far ahead of the car the target waypoint should be.
+        state: The current state of the vehicle.
+        error_state: The error between the vehicles current state and the target state.
+
+
+    """
+
+    def __init__(self):
+        """initialize the Path planning Node.
+
+        Initialize the path planning node, and set up the publishers / subscribers.
+
+        """
+        super().__init__("waypoints_path_planner_node")
+
+        # update frequency of this node
+        self.freq = 10.0
+
+        self.file = open(
+            "/home/art/art/workspace/src/path_planning/waypoints_path_planner/waypoints_path_planner/path.csv"
+        )
+        self.ref_traj = np.loadtxt(self.file, delimiter=",")
+        self.final_x = self.ref_traj[-1, 0]
+        self.final_y = self.ref_traj[-1, 1]
+
+        # READ IN SHARE DIRECTORY LOCATION
+        package_share_directory = get_package_share_directory("waypoints_path_planner")
+
+        self.declare_parameter("lookahead", 2.0)
+        self.lookahead = (
+            self.get_parameter("lookahead").get_parameter_value().double_value
+        )
+
+        # data that will be used by this class
+        self.state = VehicleState()
+
+        # subscribers
+        qos_profile = QoSProfile(depth=1)
+        qos_profile.history = QoSHistoryPolicy.KEEP_LAST
+        self.sub_state = self.create_subscription(
+            VehicleState, "~/input/vehicle_state", self.state_callback, qos_profile
+        )
+
+        # publishers
+        self.pub_err_state = self.create_publisher(
+            VehicleState, "~/output/error_state", 10
+        )
+        self.timer = self.create_timer(1 / self.freq, self.pub_callback)
+
+    # function to process data this class subscribes to
+    def state_callback(self, msg):
+        """The state Callback.
+
+        The callback to read the current state of the vehicle.
+
+        Args:
+            msg: The message received from the vehicle state topic
+
+        """
+        # self.get_logger().info("Received '%s'" % msg)
+        self.state = msg
+
+    def wpts_path_plan(self):
+        """The Waypoints path planner.
+
+        Computes the error state between the target state and the current state and the target state. First, get the target point which is closest ot the current state (plus the lookahead distance in the x direction). Then, get the error state by subtracting the current state from the target state (modular subtraction for the heading).
+
+        Returns:
+             The error state, and the target reference waypoint we are currently going to.
+
+        """
+        x_current = self.state.pose.position.x
+        y_current = self.state.pose.position.y
+        # theta_current = np.arctan2( 2*(x*w+z*y), x**2-w**2+y**2-z**2 )
+        theta_current = self.state.pose.orientation.z - 0.24845347641462115
```

Comment:
lol long number, what is it?

Reply:
This has been here for a while. Weird bug that we haven't done anything about yet where the LTP we are driving on and converting GPS coordinates to is rotated by a weird amount compared to the Magnetometer readings...

Reply:
We should start a separate issue about that.